From 73da2a77e53ef5b264925a7de57c6bdaeaaa005d Mon Sep 17 00:00:00 2001
From: Weifeng Liu <weifeng.liu@intel.com>
Date: Fri, 5 Jan 2024 11:14:06 +0000
Subject: [PATCH 07/14] drm/virtgpu: use DMA API when importing GEMs

When CONFIG_DMABUF_DEBUG is active, device fails to find memfd for the
buffer being imported because the addresses are managled. Work around
this by using DMA API always in PRIME import function.

Signed-off-by: Weifeng Liu <weifeng.liu@intel.com>
---
 drivers/gpu/drm/virtio/virtgpu_prime.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/virtio/virtgpu_prime.c b/drivers/gpu/drm/virtio/virtgpu_prime.c
index 655159fd138f..bddfd6e84086 100644
--- a/drivers/gpu/drm/virtio/virtgpu_prime.c
+++ b/drivers/gpu/drm/virtio/virtgpu_prime.c
@@ -170,7 +170,17 @@ static int virtio_gpu_sgt_to_mem_entry(struct virtio_gpu_device *vgdev,
 	struct scatterlist *sg;
 	int si;
 
-	bool use_dma_api = !virtio_has_dma_quirk(vgdev->vdev);
+	/* Extracting struct page from sgt mapped by DMA-BUF is explicitly
+	 * prohibited according to this patch [1]. Specifically, when
+	 * CONFIG_DMABUF_DEBUG is enabled, the address of the struct page in
+	 * sgt is mangled so that we get incorrect result from sg_phys.
+	 *
+	 * For our use case, the DMA mapping is a dumb operation performing
+	 * identical translation, thus it's safe to use DMA address instead.
+	 *
+	 * [1] https://patchwork.freedesktop.org/patch/414455/
+	 */
+	const bool use_dma_api = true;
 	if (use_dma_api)
 		*nents = table->nents;
 	else
-- 
2.17.1

